### PR TITLE
KAFKA-9667: Connect JSON serde strip trailing zeros

### PR DESCRIPTION
**Pull request by big-andy-coates**
Fixes: [KAFKA-9667](https://issues.apache.org/jira/browse/KAFKA-9667)

The Connect Json serde was recently enhanced to support serializing decimals as standard JSON numbers, (Original work done under [KAFKA-8595](https://issues.apache.org/jira/browse/KAFKA-8595)), e.g. `1.23`.  However, there is a bug in the implementation: it's stripping trailing zeros!  `1.23` is not the same as `1.230`.  The first is a number accurate to 2 decimal places, where as the later is accurate to 3 dp.

It is important that trailing zeros are not dropped when de(serializing) decimals.  For some use-cases it may be acceptable to drop the trailing zeros, but for others it definitely is not.

#### Current Functionality
If a JSON object was to contain the number `1.230` then the Java JsonDeserializer would correctly deserialize this into a `BigDecimal`. The BigDecimal would have a scale of 3, which is correct.

However, if that same BigDecimal was then serialized back to JSON using the Java JsonSerializer it would incorrectly strip the zeros, serializing to `1.23`. 

#### Expected Functionality
When serializing, trailing zeros should be maintained.  For example, a BigDecimal such as `1.230`, (3 dp), should be serialized as `1.230`. 

#### Compatibility
Both the old serialized number, e.g. `1.23`, and the proposed corrected serialized number, e.g. `1.230`, are valid JSON numbers. Downstream consumers should have no issue deserializing either.

This is not super urgent, but would be good to get into the next point releases of 2.4 and 2.5.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by big-andy-coates** [comment]
Related to https://github.com/confluentinc/ksql/issues/4710

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (agavra) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: what about exposing a `JsonNodeFactory` as an argument to make this more extensible?

**Review by agavra — COMMENTED** [review]
LGTM! (Not approving since I'm not a committer)

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (rhauch) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: Why is `false` (inexact decimals) the default for the no-arg constructor? If this is an attempt to maintain backward compatibility, we should consider whether this bug, when fixed, compatible. Seems like it would be, since having the deserializer use the trailing zeros would be fine/better than not using them.

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: Can do, but is there any point for a package local constructor? We can change as we need to anyway, right?

Comment (agavra) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: 😈 I "have a friend" who sometimes puts new classes in the same package so that they can use package private methods. This friend understands that such code is brittle.

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (rhauch) [inline connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: Passing in a `JsonNodeFactory` instance might be better here, because then the `JsonConverter` instances and `JsonDeserializer` in each converter will all use the same `JsonNodeFactory`.

As it currently stands, the `JsonDeserializer` has it's own instance of the `JsonNodeFactory`, and it's possible that they could be set up differently in the future and not caught.

**Review by rhauch — COMMENTED** [review]
Thanks, @big-andy-coates. The changes look great, but I have a few comments in-line.

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: Happy to flip it. 

Do we want to make the other constructor public so that others can control these things?

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: On it...

Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: What do you think about also enabling `SerializationFeature.WRITE_BIGDECIMAL_AS_PLAIN`?

This would mean serialiser won't write numbers using scientific notation, e.g. `100` , to `1E+2`.  Both are valid JSON, and the later takes up less space, but can confuse some users and maybe some downstream systems.

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: lol. I don't know what you're talking about!

Changed...

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonDeserializer.java]: Flipped

File: connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java
Comment (big-andy-coates) [reply connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java]: Discussion offline - not going with none SN as this woudl require a KIP

**Comment by big-andy-coates** [comment]
Waiting on code freeze to lift before merging.

**Comment by big-andy-coates** [comment]
@rhauch can this be merged now?

**Comment by big-andy-coates** [comment]
@rhauch can this be merged now?

**Comment by rhauch** [comment]
okay to test

**Review by rhauch — APPROVED** [review]
LGTM. Thanks, @big-andy-coates.

Will merge pending a green build.

**Comment by rhauch** [comment]
retest this please

**Pull request merged**
